### PR TITLE
Use JSON schema to validate webpack config (fixes #2971)

### LINKED ISSUE #2971: Use JSON schema to validate webpack configuration
**Current behavior:**

`webpack` constructor accepts any object and it will silently fail when it does not recognise a property, e.g.

```js
const webpackConfig = {
  devtool: 'source-map',
  debug: true,
  context: __dirname,
  target: 'node',
  entry: {
    'app': [
      path.resolve(__dirname, './src')
    ]
  },
  pathinfo: true,
  output: {
    path: outputDir,
    filename: '[name].js'
  }
};
```

Looking at the above configuration, the error is not immediately obvious.

**Expected/desired behavior:**

Use http://json-schema.org/ to enforce `webpack` configuration.

In the above example, an error would be raised since `pathinfo` property belongs to the `output` object.

**What is the motivation / use case for changing the behavior?**

To prevent unintentional configuration silently failing.

### PR DESCRIPTION
**Pull request by gajus**
**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

- [x] Feature

**What is the current behavior?** (You can also link to an open issue here)

Validation is limited to a simple condition:

```js
if(!options.entry && !options.plugins) {
  throw new Error("Passed 'options' object does not look like a valid webpack configuration");
}
```

**What is the new behavior?**

Validation logic is expressed using `./schemas/webpackOptionsSchema.json` JSON schema.

**Does this PR introduce a breaking change?**

- [x] Yes

If this PR contains a breaking change, please describe the following... 

* Impact:
    
    An error is thrown when an unknown configuration is encountered in webpack options object.
    This error might appear now because previously unknown configuration was ignored.
    
* Migration path for existing applications:
    
    Remove unknown configuration properties from webpack options object.
    To configure a loader, use `LoaderOptionsPlugin`.
    For a list of valid configuration properties, refer to the webpack documentation (https://webpack.github.io/docs/configuration.html).

* Github Issue(s) this is regarding:
    
    https://github.com/webpack/webpack/issues/2971

**Other information**:

* Need to discuss how to approach testing.
    
    JSON schema is declarative. Therefore, the contents of the schema should not be tested. The implementation testing is left to https://github.com/epoberezkin/ajv.
* Need to discuss how to inform user about errors.
    
    Just throwing an error is not enough.
    > Passed 'options' object does not look like a valid webpack configuration
    
    I am fine with the current implementation – throwing an error and printing detail info to the `console.log`. I can see how someone can dislike this.

### Outstanding tasks

* [ ] Waiting for @sokra to come back on https://github.com/webpack/webpack/pull/2974#discussion_r78160976
* [x] Waiting for a decision/ solution regarding a pre-existing failing test condition https://github.com/webpack/webpack/pull/2974#issuecomment-245860191

### CODE REVIEW

File: schemas/webpackOptionsSchema.json [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+          	{
+	            "$ref": "#/definitions/common.loaders"
+	          }
+          ],
+          "description": "An array of automatically applied loaders."
+        }
+      },
+      "type": "object"
+    },
+    "output": {
```
Comment (TheLarkInn) [inline schemas/webpackOptionsSchema.json]: Missing auxiliaryComments, heres the PR for it. 

I hadn't gotten to documenting it yet. :eyes: 

https://github.com/webpack/webpack/pull/2113

It is essentially: 
`auxiliaryComments: string|object`

If `object` properties are: 
```
{
  root?: string
  commonjs?: string
  commonjs2?: string
  amd?: string
}
```

Comment (gajus) [inline schemas/webpackOptionsSchema.json]: f94e9457f61289cd076c8262ed2d1bdee37fbcfb

**Comment by gajus** [comment]
Tests are failing because:

```js
[ { keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'name' },
    message: 'should NOT have additional properties' } ]
```

```js
[ { keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'optimize' },
    message: 'should NOT have additional properties' } ]
```

```js
[ { keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'worker' },
    message: 'should NOT have additional properties' } ]
```

```js
[ { keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'updateIndex' },
    message: 'should NOT have additional properties' } ]
```

```js
[ { keyword: 'additionalProperties',
    dataPath: '',
    schemaPath: '#/additionalProperties',
    params: { additionalProperty: 'stats' },
    message: 'should NOT have additional properties' } ]
```

can anyone contribute a description for these? (they are not present in https://webpack.github.io/docs/configuration.html)

File: lib/webpack.js [hunk 2]
```diff
@@ -15,7 +16,9 @@ function webpack(options, callback) {
 			return webpack(options);
 		}));
 	} else if(typeof options === "object") {
-		if(!options.entry && !options.plugins) {
+		var webpackOptionsValidationErrors = validateWebpackOptions(options);
+		if(webpackOptionsValidationErrors.length) {
+			console.log("'options' validation errors:", webpackOptionsValidationErrors);
```
Comment (sokra) [inline lib/webpack.js]: You are only allowed use console.log in the CLI not in the API.

Instead rewrite the code to throw a CustomError with a `validationErrors` property. Add code to the CLI to check for this property and format the errors.

**Comment by sokra** [comment]
Please add `*.json` to the `.editorconfig`.

**Comment by sokra** [comment]
`name` is a valid property used to give configs a name when using multiple configurations in an array.

`optimize` is invalid. It's old syntax and no longer used.

`worker` is used by the `worker-loader`, but it should not be on the config object, but the config should use the LoadersOptionsPlugin instead.

`stats` is valid. It allows to pass stat options to the CLI.

`updateIndex` is used by the HMR test cases, but can also be rewritten to use the LoadersOptionsPlugin

**Comment by TheLarkInn** [comment]
Info on stats @gajus can be gathered from here.

https://github.com/webpack/webpack/blob/master/lib/Stats.js

**Comment by TheLarkInn** [comment]
Also would this error for a user if they added an additional custom property (some sort of metadata) to their config? I know quite a few who do this, and use those options, etc.

**Comment by jhnns** [comment]
I'm not sure if we should throw when additional keys are encountered... webpack 2 should not have too many breaking changes. It's ok to discourage it and display warnings, but I don't think that we should break. The sass- and the less-loader for instance provide the possibility to read options from the config (because we had to deal with functions as options).

**Comment by gajus** [comment]
> Please add *.json to the .editorconfig.

b4e6604

File: lib/webpack.js (hunk 2 above)
Comment (gajus) [inline lib/webpack.js]: What about users that use webpack programmatically? e.g. most of the React applications will have a setup using `webpack` and `webpack-dev-middleware`.

Do I need to document presence of the `validationErrors` property?

Comment (sokra) [inline lib/webpack.js]: We will document it in the new documentation.

https://github.com/webpack/webpack.js.org/blob/develop/content/api/node.md

**Comment by gajus** [comment]
> worker is used by the worker-loader, but it should not be on the config object, but the config should use the LoadersOptionsPlugin instead.

This ought to be a simple change from:

```js
module.exports = {
	worker: {
		output: {
			filename: "hash.worker.js",
			chunkFilename: "[id].hash.worker.js"
		}
	}
}
```

to:

```js
var webpack = require("../../");
module.exports = {
	plugins: [
		new webpack.LoaderOptionsPlugin({
			options: {
				worker: {
					output: {
						filename: "hash.worker.js",
						chunkFilename: "[id].hash.worker.js"
					}
				}
			}
		})
	]
}
```

Correct?

**Comment by gajus** [comment]
> Correct?

Yes.

I forgot `npm link webpack`.

**Comment by gajus** [comment]
@sokra 

```bash
git clone https://github.com/webpack/webpack.git
cd webpack
npm install
npm link
npm link webpack
npm test
```

is giving:

```
  1) ConfigTestCases extract-text issue-14 should compile:
     Error: global leak detected: addresses
      at Runner.checkGlobals (/dev/node_modules/mocha/lib/runner.js:210:21)
      at Runner.<anonymous> (/dev/node_modules/mocha/lib/runner.js:73:10)
      at emitOne (events.js:101:20)
      at Runner.emit (events.js:188:7)
      at /dev/node_modules/mocha/lib/runner.js:557:14
      at done (/dev/node_modules/mocha/lib/runnable.js:287:5)
      at /dev/node_modules/mocha/lib/runnable.js:360:7
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)
```

Unrelated to this PR.

**Comment by gajus** [comment]
> worker is used by the worker-loader, but it should not be on the config object, but the config should use the LoadersOptionsPlugin instead.

70e4a1e

> updateIndex is used by the HMR test cases, but can also be rewritten to use the LoadersOptionsPlugin

e42ddb2

File: .editorconfig [hunk 3]
```diff
@@ -1,5 +1,5 @@
 root = true
 
-[*.js]
+[*.{js,json}]
```
Comment (sokra) [inline .editorconfig]: for npm reasons `package.json` must be 2 spaces indent

File: lib/webpack.js [hunk 4]
```diff
@@ -15,8 +16,11 @@ function webpack(options, callback) {
 			return webpack(options);
 		}));
 	} else if(typeof options === "object") {
-		if(!options.entry && !options.plugins) {
-			throw new Error("Passed 'options' object does not look like a valid webpack configuration");
+		var webpackOptionsValidationErrors = validateWebpackOptions(options);
+		if(webpackOptionsValidationErrors.length) {
+			var validationError = new Error("Passed 'options' object does not look like a valid webpack configuration");
```
Comment (sokra) [inline lib/webpack.js]: Create a custom Error class.

example: https://github.com/webpack/webpack/blob/master/lib/CaseSensitiveModulesWarning.js

File: bin/convert-argv.js [hunk 5]
```diff
@@ -553,5 +554,11 @@ module.exports = function(optimist, argv, convertOptions) {
 			console.error("Use --help to display the CLI options.");
 			process.exit(-1); // eslint-disable-line
 		}
+		var webpackOptionsValidationErrors = validateWebpackOptions(options);
+
+		if (webpackOptionsValidationErrors.length) {
+			console.error("Configuration is invalid. Fix the following errors:", webpackOptionsValidationErrors);
+			process.exit(-1); // eslint-disable-line
+		}
```
Comment (sokra) [inline bin/convert-argv.js]: remove this block and catch the exceptions thrown by the API in `bin/webpack.js` instead.

File: .editorconfig (hunk 3 above)
Comment (gajus) [inline .editorconfig]: 1d4869a

Technically, we could have a separate rule for `./package.json` and the rest.

**Comment by jhnns** [comment]
No comments on the error handling?

File: bin/convert-argv.js (hunk 5 above)
Comment (gajus) [inline bin/convert-argv.js]: Have you seen the comment about this?

https://github.com/webpack/webpack/pull/2974/commits/efc4113e12b3e1677f4f1b4bc54899b8f3abb8fb

Please advise if to proceed in your direction after you have read it.

**Comment by gajus** [comment]
> The sass- and the less-loader for instance provide the possibility to read options from the config (because we had to deal with functions as options).

My understanding is that all options need to be passing using `webpack.LoaderOptionsPlugin`.

**Comment by jhnns** [comment]
There will still be loader queries in webpack 2 (and tbh I don't know what the difference between an option and a query is).

My problem with this change is, that it needs some time to be implemented in the community. Loader authors need to update their code and publish new versions of it, people need to update their dependencies. I would like to have a smooth transition where you get a warning with webpack 2 and an error with webpack 3.

**Comment by gajus** [comment]
Options and queries are interchangeable.

Loader authors are not affected. No code change is required to the loader to accept options passed using LoaderOptionsPlugin. Therefore, no update is needed of dependencies.

As I said earlier, this only affects webpack configuration.

**Comment by jhnns** [comment]
> No code change is required to the loader to accept options passed using LoaderOptionsPlugin. Therefore, no update is needed of dependencies.

It depends on how robust the loader was developed. I took a quick look at the sass- and the less-loader which both had to read from the webpack config object because they had to support function options and it seems like they all support the new behavior by accident because they populate the config from the query *and* the webpack config object (see [here](https://github.com/webpack/less-loader/blob/7ab6a2ef7d17e1ebaa9d15e8cf79fc0120c21523/index.js#L46-L52) and [here](https://github.com/jtangelder/sass-loader/blob/master/index.js#L386-L394)). Also the [markdown-loader](https://github.com/peerigon/markdown-loader/blob/1454d5152a5ccb172213faadfe034d1bb75e1bd2/index.js#L23) does this by accident. If this code was written in a way these specific options can **only** be configured on the webpack options object, the author would need to update it.

Since the sass- and the less-loader would support the new way out-of-the-box, the impact of this change is probably not as big as I've assumed. However, I don't know how many loaders out there would need a fix for this.

@gajus I really like this change and thanks for your effort 👍. I just don't want to introduce too many *actual* breaking changes just because it's webpack 2. We may break things, but l want to give people time to adjust their code.

What do other people think about this? @TheLarkInn @bebraw @sokra

**Comment by gajus** [comment]
> ConfigTestCases extract-text issue-14 should compile:
> Error: global leak detected: addresses

Thank you @sokra 

37066fc692a0e981f25aeb6010bedf5e5f886c53

**Comment by bebraw** [comment]
@jhnns Warnings in webpack 2 and failing hard in webpack 3 would seem the sanest way to me too. That way we get at least some value from validation while giving depending projects time to catch up.

**Comment by gajus** [comment]
> @jhnns Warnings in webpack 2 and failing hard in webpack 3 would seem the sanest way to me too. That way we get at least some value from validation while giving depending projects time to catch up.

Sounds reasonable.

I would only argue that since webpack 2 is still not released, making this a breaking change shouldn't be a concern. This is not changing webpack configuration API. It only enforces not to use configuration that shouldn't be there in the first place. Considering that any existing project that uses webpack@2-beta at the moment is under an active development, it would be a minor annoyance.

**Comment by TheLarkInn** [comment]
It definitely needs to *only warn* in webpack 2. We are only holding webpack 2 in beta until documentation is released. This is essentially a Release Candidate. This features original purpose is to help validate the types of properties that are in the schema. 

If these can be handled by giving deprecation warnings, please do so. After webpack 2, v3 brings a lot of breaking changes for usability .etc.

**Comment by gajus** [comment]
> It definitely needs to only warn in webpack 2. We are only holding webpack 2 in beta until documentation is released. This is essentially a Release Candidate. This features original purpose is to help validate the types of properties that are in the schema.

Understood.

Earlier @sokra comment asked to remove `console.log` from the API. If we are going this route, I will need to re-add it.

**Comment by jhnns** [comment]
> Earlier @sokra comment asked to remove console.log from the API

You're right 😁.  @sokra is probably more keen to break things, but I don't agree with him in this regard. Since webpack is one of the most popular JS projects, we should be more conservative with breaking things.

**Comment by jhnns** [comment]
> It only enforces not to use configuration that shouldn't be there in the first place

Unfortunately, this is not true. [We have added this as advice to our docs](http://webpack.github.io/docs/how-to-write-a-loader.html#programmable-objects-as-query-option) as a makeshift because we did not know how to deal with this.

**Comment by TheLarkInn** [comment]
@andreypopp you have dealt with some of the error messaging. Do you have a good suggestion for @gajus that would allow for warning on additional properties in the webpack config?

**Comment by TheLarkInn** [comment]
Alright after @webpack/collab-team discussed, we decided that we _will let this error_, however to merge this PR must fulfill these requirements:

- [ ] Provides error messages that describe in *clear detail* to the user how to remedy this error via LoaderOptionsPlugin();. Since the validation knows exactly what properties are additional, you can use the erroring data to output a solution the user can copy and paste into their config. 

- [ ] Modify your Breaking Change instructions in PR Long message field to be more than:

> Migration path for existing applications: Fix configuration.

User will need to see use case and example of what will error, and how to remedy this.

**Comment by gajus** [comment]
> [..] Since the validation knows exactly what properties are additional, you can use the erroring data to output a solution the user can copy and paste into their config.

That makes an assumption that the additional property is a loader property.

There is no way to distinguish whether user made a typo (e.g. "debuq" instead of "debug"), or if the user's intention was to pass a property to the loader (e.g. "sassLoader").

Therefore, telling user to add the configuration to `LoaderOptionsPlugin` would be misguiding.

However, a phrasing such as:

> "foo" is an unknown property.
> To configure a loader, use `LoaderOptionsPlugin`.
> For everything else, refer to the webpack documentation (https://webpack.github.io/docs/configuration.html).

**Comment by gajus** [comment]
> Modify your Breaking Change instructions in PR Long message field to be more than:

How about:

> Remove unknown configuration properties from webpack options object.
> To configure a loader, use `LoaderOptionsPlugin`.
> For a list of valid configuration properties, refer to the webpack documentation (https://webpack.github.io/docs/configuration.html).

**Comment by sokra** [comment]
yeah, but the link should be http://webpack.js.org/api/configuration/

Don't link to the old documentation.

**Comment by TheLarkInn** [comment]
> Using custom properties which do not exist on the webpack schema will now throw an error during validation. Users now will need to use LoaderOptionsPlugin({}). Therefore the following example: 

```javascript
module.exports = config;

const config = {
  sassLoader: {
    includePaths: [
      /*...*/
    ]
  }

};
```

Will need to be migrated to look as follows: 

```javascript
module.exports = config;

const customLoaderOptions = {
  sassLoader: {
      includePaths: [
        /*...*/
      ]
    }
};

const config = {
  plugins: [
    webpack.LoaderOptionsPlugin({
      options: customLoaderOptions;
    });
  ]

};
```

Something like this I would assume. In addition providing the link to documentation.

**Comment by TheLarkInn** [comment]
Also tagging @gaearon @hansl @filipesilva @taylorotwell as this could impact their respective projects if they are using loaders which require custom properties in the webpack config. I'll try and communicate this in your respective repo's also.

**Comment by sokra** [comment]
btw. the long-term goal is to get these options back into the query object, by allowing any value in queries...

**Comment by TheLarkInn** [comment]
> That makes an assumption that the additional property is a loader property.

> There is no way to distinguish whether user made a typo (e.g. "debuq" instead of "debug"), or if the user's intention was to pass a property to the loader (e.g. "sassLoader").

This is true. I would want the error to still be intuitive for the user if it can be. Can we do better than what we have?

**Comment by gajus** [comment]
Lets start with the error message.

    this.message = "Passed 'options' object does not look like a valid webpack configuration.";

This can be rephrased to:

> Invalid configuration object. Webpack has been initialised using a configuration object that does not match API schema. Inspect validationErrors property of the error to learn what checks have failed.

Regarding the console log error.

    console.error("Configuration is invalid. Fix the following errors:", webpackOptionsValidationErrors);

This can be rephrased to:

> Invalid configuration object. Webpack has been initialised using a configuration object that does not match API schema. The following checks have failed.
> [... list of failed checks]

**Comment by jhnns** [comment]
Sounds good to me. 👍 

Shouldn't it be "that does not match **the** API schema"?

**Comment by gajus** [comment]
> Shouldn't it be "that does not match the API schema"?

Corrected. Thank you.

**Comment by gajus** [comment]
Message phrasing has been improved. 9e0a95e

Can this be merged?

**Comment by rupesh1** [comment]
If I do an npm install with a package.json setup like: `"webpack": "https://github.com/webpack/webpack"`

I get the error:
```
> cross-env NODE_ENV=production webpack --bail --progress --color --display-error-details --profile

module.js:327
    throw err;
    ^

Error: Cannot find module '../schemas/webpackOptionsSchema.json'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/rupesh/Documents/src/project-name/node_modules/webpack/lib/validateWebpackOptions.js:5:28)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
```

And sure enough the folder and file in the path mentioned are missing on the file system. 
Is anyone else seeing this?

**Comment by sokra** [comment]
schemas is missing in `package.json` `files`. You can send a PR.

**Comment by gajus** [comment]
> schemas is missing in package.json files. You can send a PR.

https://github.com/webpack/webpack/pull/3013

**Comment by MoOx** [comment]
A breaking change from 2.1.0-beta22 to 2.1.0-beta23 really sucks. You might really broke a lot of CI system today...

**Comment by ThatTomPerson** [comment]
beta22 and beta23 are both prerelease versions and there has not been a 2.x release yet.
sokra has every right as defined by semver to make breaking changes

**Comment by MoOx** [comment]
Having the "right" to do something is a thing. Not caring about current users (that help to test prerelease right?) and upgrade path is another thing.

**Comment by ThatTomPerson** [comment]
"testing" and "depending" on prerelease software are different things
if webpack can't make breaking changes then it will just end up with tonnes of bloat and be slow

I think this is a good change, random plugins should not pollute the config with random keys in case webpack wants to use them in the future

instead of complaining about it, we could help come up with a solution, for example,
maybe webpack could add a `loaderConfig: {}` option which loader and plugin authors could use for their config without using the global config

so loaders just go `config.loaderConfig[loaderName]` instead of `config[loaderName]`

Also thanks, I have been looking into static site generators atm so I might look at phenomic when I get home

**Comment by MoOx** [comment]
> I think this is a good change, random plugins should not pollute the config with random keys in case webpack wants to use them in the future

It definitely is.

I think a simpler workaround than having to use a plugin to pass data should be offered. The errow thrown does not event mention a solution. I could say

```
If your configuration currently use non webpack related properties (eg: eslint, sass, postcss etc.), you should pass those data by using `LoaderOptionsPlugin`. Please refer to XXX for more detail
```

I think the best solution will be to support real JS for query. It's the best user experience we can offer an make configuration for loader even more clear than having to split in multiple places.

**Comment by huizi666** [comment]
i have a trouble in webpack2 . but i'm tring to use webpack1' way to code my webapck2.
and ....there is some thing wrong in postcss-loader 
`<addr>` module : {
		loaders : [
		{test:/\.js$/,loader:'babel-loader',exclude:path.resolve(__dirname,'node_modules'),
		//绝对路径 只解析src下的提高打包性能
		include:path.resolve(__dirname,'src')},
		{test:/\.css$/,loader:'style-loader!css-loader!postcss-loader'}
		]
	},
	plugins : {
		//for postcss
		new webpackLoaderOptionsPlugin({
			options : {
				postcss : function(){
					return [auroprefixer]
				},
			    devServer: {
			        contentBase: "./public", //本地服务器所加载的页面所在的目录
			        colors: true, //终端中输出结果为彩色
			        historyApiFallback: true, //不跳转
			        inline: true //实时刷新
			      }	
			}
		}),
======just part of them

there is error
angular-oms-front\webpack.config.js:27
                new webpackLoaderOptionsPlugin({
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^
SyntaxError: Unexpected identifier
    at Object.exports.runInThisContext (vm.js:76:16)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at requireConfig (E:\another\angular-oms-front\node_modules\webpack\bin\convert-argv.js:96:18)
    at E:\another\angular-oms-front\node_modules\webpack\bin\convert-argv.js:109:17


(i'm not good at english .waiting your reply)

**Comment by bebraw** [comment]
@huizi666 `plugins` should be an array. Could you please open support request at Stack Overflow and other channels? Thanks. 👍

**Comment by huizi666** [comment]
I dont understand about "support request",but this is my git ，could you please gitclone git@github.com:huizi666/angular-webpack.git  . And checkout the webpack.config.js   Thanks~~ @bebraw

**Comment by huizi666** [comment]
@bebraw  may you can send me an email : superhuizi0716@gmail.com ~ ^_^

**Pull request merged**